pkg/collector/utils.go

```diff
@@ -20,34 +20,44 @@ import (
 	"github.com/sustainable-computing-io/kepler/pkg/cgroup"
 	collector_metric "github.com/sustainable-computing-io/kepler/pkg/collector/metric"
 	"k8s.io/klog/v2"
+
+	"github.com/sustainable-computing-io/kepler/pkg/kubernetes"
 )
 
+// this function is only called with the watcher delayed to sync and update the container info or if the watcher is not enabled
 func (c *Collector) createContainersMetricsIfNotExist(containerID string, cGroupID, pid uint64, withCGroupID bool) {
 	if _, ok := c.ContainersMetrics[containerID]; !ok {
-		var err error
-		var podName, containerName string
-
-		// the acceletator package does not get the cgroup ID, then we need to verify if we can call the cgroup with cGroupID or not
-		podName, _ = cgroup.GetPodName(cGroupID, pid, withCGroupID)
-		containerName, _ = cgroup.GetContainerName(cGroupID, pid, withCGroupID)
-
+		// We feel the info with generic values because the watcher will eventually update it.
+		podName := c.systemProcessName
+		containerName := c.systemProcessName
 		namespace := c.systemProcessNamespace
 
-		if containerName == c.systemProcessName {
-			// if the systemProcess already exist in ContainersMetrics do not overwrite the data
-			if _, exist := c.ContainersMetrics[containerID]; exist {
-				return
-			}
-			containerID = c.systemProcessName
-		} else {
-			namespace, err = cgroup.GetPodNameSpace(cGroupID, pid, withCGroupID)
-			if err != nil {
-				klog.V(5).Infof("failed to find namespace for cGroup ID %v: %v", cGroupID, err)
-				namespace = "unknown"
+		// In case the pod watcher is not enabled, we need to retrieve the information about the
+		// pod and container from the kubelet API. However, we prefer to use the watcher approach
+		// as accessing the kubelet API might be restricted in certain systems.
+		// Additionally, the code that fetches the information from the kubelet API utilizes cache
+		// for performance reasons. Therefore, if the kubelet API delay the information of the
+		// containerID (which occasionally occurs), the container will be wrongly identified for its entire lifetime.
+		if !kubernetes.IsWatcherEnabled {
+			podName, _ = cgroup.GetPodName(cGroupID, pid, withCGroupID)
+			containerName, _ = cgroup.GetContainerName(cGroupID, pid, withCGroupID)
+			if containerName == c.systemProcessName {
```

Comment:
i suspect there still could be a chance of split brain. If kubelet cannot detect pod name at this time, apiserver won't be able to do it either.

Reply:
In scenarios where it is necessary, please note that you can disable the apiserver by using the flag `-apiserver=false`.